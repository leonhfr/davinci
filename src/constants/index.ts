// Definition.
export const FLICKR_API_ENDPOINT = 'https://api.flickr.com/services/rest/';

export const FLICKR_API_PARAMETERS = { format: 'json' };
