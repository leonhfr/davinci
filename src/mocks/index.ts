// Internal.
import { CreateResponseInput } from '../lib/Response';
import * as Types from '../types';

// Code.
export const createMockResponseInput = (
  response?: any
): CreateResponseInput => {
  const _response = response || {};
  return {
    page: _response.page || 1,
    pages: _response.pages || 25225,
    perpage: _response.perpage || 2,
    total: _response.total || 50450,
    photo: _response.photo || [],
  };
};

export const createMockResponse = (response?: any) => {
  const _response = response || {};
  return {
    page: _response.page || 1,
    pages: _response.pages || '25225',
    perpage: _response.perpage || 2,
    total: _response.total || '50450',
    photo: _response.photo || [],
  };
};

export const createMockPhoto = (photo?: any) => {
  const _photo = photo || {};
  return {
    id: _photo.id || '32457360587',
    owner: _photo.owner || '41710899@N08',
    secret: _photo.secret || '69b27ceb18',
    server: _photo.server || '4898',
    farm: _photo.farm || 5,
    title: _photo.title || 'Arc de Triomf',
    ispublic: 1,
    isfriend: 0,
    isfamily: 0,
    description: _photo.description || {
      _content:
        'The Arc de Triomf was built as the entrance of the 1888 Barcelona Universal Exposition. It was designed by architect Josep Vilaseca i Casanovas in the Neo-Mudéjar style.',
    },
    ownername: _photo.ownername || 'jdf_92',
    views: _photo.views || '16',
    tags:
      _photo.tags ||
      'spain barcelona arcdetriomf 1888barcelonaworldfair neomudéjar 1888barcelonauniversalexposition catelonia',
    latitude: _photo.latitude || '41.390791',
    longitude: _photo.longitude || '2.18114',
    accuracy: 16,
    context: _photo.context || 0,
    place_id: 'iFsnXNZTUriu3usbkw',
    woeid: '20220092',
    geo_is_family: 0,
    geo_is_friend: 0,
    geo_is_contact: 0,
    geo_is_public: 1,
  };
};

export const photoProperties: Types.PhotoProperties = {
  zoneId: '4ab7068b-6c6c-46d2-8009-1d7d1ab35a3b',
  inside: false,
};

export const apiKey = 'test_api_key';

export const searchOptions: Types.SearchOptions = {
  bbox: [2.18, 41.4, 2.19, 41.5],
  min_upload_date: 1546300800,
  max_upload_date: 1546387200,
  tags: ['foo', 'bar'],
  page: undefined,
};
