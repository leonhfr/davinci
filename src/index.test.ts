// Packages.
import axios from 'axios';

// Internal.
import { FLICKR_API_ENDPOINT } from './constants';
import * as index from './index';
import { Flickr } from './index';
import * as Mocks from './mocks';

// Code.
describe('index', () => {
  it('should expose the expected entries', () => {
    expect(index).toMatchSnapshot();
  });
});

describe('Flickr', () => {
  describe('API', () => {
    let axiosGetSpy: jest.SpyInstance;
    beforeEach(() => {
      axiosGetSpy = jest.spyOn(axios, 'get');
    });
    afterEach(() => {
      jest.restoreAllMocks();
    });

    it('should be defined', () => {
      expect(Flickr.API).toBeDefined();
    });

    describe('search', () => {
      it('should be defined', () => {
        expect(Flickr.API.search).toBeDefined();
      });
      it('should return an error if the request fails', async () => {
        const mockApiResponse = {
          data: {},
          status: 500,
          statusText: 'ERROR',
          headers: '',
          config: '',
        };
        axiosGetSpy = jest
          .spyOn(axios, 'get')
          //@ts-ignore
          .mockImplementation(async () => Promise.resolve(mockApiResponse));

        const result = await Flickr.API.search(
          Mocks.apiKey,
          Mocks.searchOptions,
          Mocks.photoProperties
        );
        expect(result).toMatchSnapshot();
      });
      it('should query the api with axios.get and return the expected result', async () => {
        const mockApiResponse = {
          data: {
            photos: Mocks.createMockResponse({
              photo: [Mocks.createMockPhoto()],
            }),
          },
          status: 200,
          statusText: 'OK',
          headers: '',
          config: '',
        };
        axiosGetSpy = jest
          .spyOn(axios, 'get')
          //@ts-ignore
          .mockImplementation(async () => Promise.resolve(mockApiResponse));

        const result = await Flickr.API.search(
          Mocks.apiKey,
          Mocks.searchOptions,
          Mocks.photoProperties
        );
        expect(result).toMatchSnapshot();
        expect(axiosGetSpy).toHaveBeenCalledTimes(1);
        expect(axiosGetSpy).toHaveBeenCalledWith(FLICKR_API_ENDPOINT, {
          api_key: 'test_api_key',
          format: 'json',
          method: 'flickr.photos.search',
          sort: 'date-posted-asc',
          accuracy: '16',
          content_type: '1',
          media: 'photos',
          extras: 'geo,description,owner_name,tags,views,o_dims',
          per_page: '250',
          bbox: '2.18,41.4,2.19,41.5',
          min_upload_date: '1546300800',
          max_upload_date: '1546387200',
          tags: 'foo,bar',
        });
      });
    });
  });
});
