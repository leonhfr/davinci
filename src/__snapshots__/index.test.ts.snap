// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flickr API search should query the api with axios.get and return the expected result 1`] = `
Response {
  "page": 1,
  "pages": 25225,
  "perpage": 2,
  "photo": Array [
    Photo {
      "context": 0,
      "description": "The Arc de Triomf was built as the entrance of the 1888 Barcelona Universal Exposition. It was designed by architect Josep Vilaseca i Casanovas in the Neo-Mudéjar style.",
      "farm": 5,
      "id": "32457360587",
      "inside": false,
      "latitude": 41.390791,
      "longitude": 2.18114,
      "owner": "41710899@N08",
      "ownername": "jdf_92",
      "secret": "69b27ceb18",
      "server": "4898",
      "tags": Array [
        "spain",
        "barcelona",
        "arcdetriomf",
        "1888barcelonaworldfair",
        "neomudéjar",
        "1888barcelonauniversalexposition",
        "catelonia",
      ],
      "title": "Arc de Triomf",
      "views": 16,
      "zoneId": "4ab7068b-6c6c-46d2-8009-1d7d1ab35a3b",
    },
  ],
  "total": 50450,
}
`;

exports[`Flickr API search should return an error if parsing fails 1`] = `[SyntaxError: Unexpected token h in JSON at position 1]`;

exports[`Flickr API search should return an error if the request fails 1`] = `[Error: Flickr API answered with 500: ERROR]`;

exports[`index should expose the expected entries 1`] = `
Object {
  "Flickr": Object {
    "API": [Function],
  },
}
`;
